docs/concurrency-adoption-guidelines.md

```diff
@@ -0,0 +1,201 @@
+# Swift Concurrency adoption guidelines for Swift Server Libraries
+
+This writeup attempts to provide a set of guidelines to follow by authors of server-side Swift libraries. Specifically a lot of the discussion here revolves around what to do about existing APIs and libraries making extensive use of Swift NIO’s `EventLoopFuture` and related types.
+
+Swift Concurrency is a multi-year effort. It is very valuable for the server community to participate in this multi-year adoption of the concurrency features, one by one, and provide feedback while doing so. As such, we should not hold off adopting concurrency features until Swift 6 as we may miss valuable opportunity to improve the concurrency model.
+
+In 2021 we saw structured concurrency and actors arrive with Swift 5.5. Now is a great time to provide APIs using those primitives. In the future we will see fully checked Swift concurrency. This will come with breaking changes. For this reason adopting the new concurrency features can be split into two phases.
+
+
+## What you can do right now
+
+### `#if` guarding code using Concurrency
+
+In order to have code using concurrency along with code not using concurrency, you may have to `#if` guard certain pieces of code. The correct way to do so is the following:
+
+```swift
+#if compiler(>=5.5) && canImport(_Concurrency)
+...
+#endif
+```
+
+Please note that you do _not_ need to _import_ the `_Concurrency` at all, if it is present it is imported automatically.
+
+```swift
+// DO NOT DO THIS.
+// Instead don't do any import and it'll import automatically when possible.
+#if compiler(>=5.5) && canImport(_Concurrency)
+import _Concurrency
+#endif
+```
+
+
+
+### API Design
+
+Firstly, existing libraries should strive to add `async` functions where possible to their user-facing “surface” APIs in addition to existing `*Future` based APIs wherever possible. These additive APIs can be gated on the Swift version and can be added without breaking existing users' code, for example like this:
+
+```swift
+extension Worker {
+  func work() -> EventLoopFuture<Value> { ... }
+  
+  #if swift(>=5.5)
+  func work() async throws -> Value { ... }
+  #endif
+}
+```
+
+If a function cannot fail but was using futures before, it should not include the `throws` keyword in its new incarnation. 
+
+Such adoption can begin immediately, and should not cause any issues to existing users of existing libraries. 
```

Comment:
I think there are broader changes that libraries can make as well, e.g. `AsyncSequence` based APIs rather than delegate based APIs, for example.

It may also be worth calling out that given we're on the road to Swift 6, libraries and the ecosystem as a whole are going to have to go through major version bumps so adding async-await APIs now can help lessen the pain of going through a major version change later because new APIs can be adopted incrementally.